capa/ghidra/capa_explorer.py

```diff
@@ -0,0 +1,245 @@
+            lib_str = Namespace.DELIMITER + "lib" + Namespace.DELIMITER
+            self.namespace = self.namespace + lib_str + self.capability.replace(" ", "-")
+        cmd = CreateNamespacesCmd(self.namespace, SourceType.USER_DEFINED)
+        cmd.applyTo(currentProgram())  # type: ignore [name-defined] # noqa: F821
+
+        return cmd.getNamespace()
+
+    def add_bookmark(self, addr, txt, category="CapaExplorer"):
+        """create bookmark at addr"""
+        currentProgram().getBookmarkManager().setBookmark(addr, "Info", category, txt)  # type: ignore [name-defined] # noqa: F821
+
+    def tag_functions(self):
+        """create function tags for capabilities"""
+
+        # self.locations[0] will always be the largest
+        # scoped offset yielded i.e. closest to an entrypoint
+        addr = toAddr(hex(self.locations[0]))  # type: ignore [name-defined] # noqa: F821
+        f = getFunctionContaining(addr)  # type: ignore [name-defined] # noqa: F821
+
+        # bookmark Mitre ATT&CK tactics @ function scope
+        if f:
+            f.addTag(self.capability)
+            if self.attack:
+                for item in self.attack:
+                    self.add_bookmark(addr, item.get("tactic"), "CapaExplorer::Mitre ATT&CK")
+
+    def bookmark_locations(self):
+        """bookmark & label findings at all scopes"""
+        st = currentProgram().getSymbolTable()  # type: ignore [name-defined] # noqa: F821
+        ns = self.create_namespace()
+
+        for addr in self.locations:
+            txt = self.capability.replace(" ", "-")
+            a = toAddr(hex(addr))  # type: ignore [name-defined] # noqa: F821
+
+            self.add_bookmark(a, txt)
+
+            # avoid re-naming functions
+            to_cont = False
+            for s in st.getSymbols(a):
+                if s.getSymbolType() == SymbolType.FUNCTION:
+                    to_cont = True
+                    txt = s.getName()
+                    createLabel(a, txt, ns, True, SourceType.USER_DEFINED)  # type: ignore [name-defined] # noqa: F821
+
+            if to_cont:
+                continue
+
+            # create labels at basic block & insn scopes
+            node_to_parse = self.node[self.locations.index(addr)]
+            if node_to_parse:
+                txt = self.recurse_node(node_to_parse)
+
+                if not txt:
+                    continue
+
+                txt = txt.replace(" ", "-")
+                createLabel(a, txt, ns, True, SourceType.USER_DEFINED)  # type: ignore [name-defined] # noqa: F821
+                continue
+
+            createLabel(a, txt, ns, True, SourceType.USER_DEFINED)  # type: ignore [name-defined] # noqa: F821
+
+
+def get_capabilities():
+    logging.basicConfig(level=logging.INFO)
+    logging.getLogger().setLevel(logging.INFO)
+
+    rules_dir: str = ""
+    try:
+        selected_dir = askDirectory("Choose capa rules directory", "Ok")  # type: ignore [name-defined] # noqa: F821
```

Comment:
is/can this be saved or will this always be asked?

Reply:
For now, it will always be asked; however, I think there's a way we can check if the project has had any saved changes associated with capa if we bump to the new Ghidrathon version.